### PR TITLE
Speedup bigtable block upload by factor of 8-10x

### PR DESCRIPTION
**Pull request by buffalu**
#### Problem
We were seeing our bigtable block upload running behind tip by several thousand blocks. We noticed it was uploading a block every ~500ms.

#### Summary of Changes
Added multiple blockstore read threads.
Run the bigtable upload in tokio::spawn context.
Run bigtable tx and tx-by-addr uploads in tokio::spawn context.

We're seeing around 50-60ms per block upload now, so we should be able to easily catch up and maintain bigtable state on the tip.

I think there's more to squeeze out, but curious what you guys think about this before I go down the next optimization rabbit hole.

### CODE REVIEW

File: ledger/src/bigtable_upload.rs [hunk 1]
```diff
@@ -126,42 +129,49 @@ pub async fn upload_confirmed_blocks(
...
 
-                    let _ = match blockstore.get_rooted_block(*slot, true) {
-                        Ok(confirmed_block) => sender.send((*slot, Some(confirmed_block))),
-                        Err(err) => {
-                            warn!(
-                                "Failed to get load confirmed block from slot {}: {:?}",
-                                slot, err
-                            );
-                            sender.send((*slot, None))
+        let (slot_tx, slot_rx) = unbounded();
```
Comment (CriesofCarrots) [inline ledger/src/bigtable_upload.rs:138]:
```suggestion
        let (slot_sender, slot_receiver) = unbounded();
```
`tx` is overloaded in solana-land ;)

File: ledger/Cargo.toml [hunk 2]
```diff
@@ -51,8 +51,8 @@ solana-transaction-status = { path = "../transaction-status", version = "=1.11.0
 solana-vote-program = { path = "../programs/vote", version = "=1.11.0" }
 tempfile = "3.3.0"
 thiserror = "1.0"
-tokio = { version = "1", features = ["full"] }
-tokio-stream = "0.1"
+tokio = { version = "1.17.0", features = ["full"] }
+tokio-stream = "0.1.8"
```
Comment (CriesofCarrots) [inline ledger/Cargo.toml:55]: Why are these changes needed?

File: ledger/src/bigtable_upload.rs [hunk 3]
```diff
@@ -18,6 +18,9 @@ use {
 // Attempt to upload this many blocks in parallel
 const NUM_BLOCKS_TO_UPLOAD_IN_PARALLEL: usize = 32;
 
+// Number of blockstore threads to read from in parallel
+const NUM_BLOCKSTORE_THREADS: usize = 16;
```
Comment (CriesofCarrots) [inline ledger/src/bigtable_upload.rs:22]: This can be a default, but we probably need to make this tunable.

**Review by CriesofCarrots — COMMENTED** [review]
Generally looks pretty good; thanks for digging into this! A couple small questions, and I'd like to play around with it a bit...

File: ledger/Cargo.toml (hunk 2 above)
Comment (buffalu) [reply ledger/Cargo.toml:55]: need that new new!

i can revert, no biggie

File: ledger/src/bigtable_upload.rs (hunk 3 above)
Comment (buffalu) [reply ledger/src/bigtable_upload.rs:22]: want me to pass in as Option<usize> to the fn with this as default if None?

File: ledger/Cargo.toml (hunk 2 above)
Comment (buffalu) [reply ledger/Cargo.toml]: fixed

File: ledger/src/bigtable_upload.rs (hunk 3 above)
Comment (CriesofCarrots) [reply ledger/src/bigtable_upload.rs:22]: Or pass in a config struct to the fn that includes blockstore_threads and blocks_to_upload_in_parallel?
I'm surprised we haven't wanted to tune `NUM_BLOCKS_TO_UPLOAD_IN_PARALLEL` more easily before now, actually

File: ledger/src/bigtable_upload.rs [hunk 4]
```diff
@@ -126,42 +129,49 @@ pub async fn upload_confirmed_blocks(
...
+            .into_iter()
+            .for_each(|b| slot_sender.send(b).unwrap());
+        drop(slot_sender);
+
+        (
+            (0..NUM_BLOCKSTORE_THREADS)
+                .map(|_| {
+                    let blockstore = blockstore.clone();
+                    let sender = sender.clone();
+                    let slot_rx = slot_receiver.clone();
```
Comment (CriesofCarrots) [inline ledger/src/bigtable_upload.rs:149]:
```suggestion
                    let slot_receiver = slot_receiver.clone();
```
Here too, please

File: ledger/src/bigtable_upload.rs (hunk 3 above)
Comment (t-nelson) [reply ledger/src/bigtable_upload.rs:22]: we typically default thread counts to a function of `num_cpus::get`

File: ledger/src/bigtable_upload.rs [hunk 5]
```diff
@@ -18,6 +18,9 @@ use {
 // Attempt to upload this many blocks in parallel
 const NUM_BLOCKS_TO_UPLOAD_IN_PARALLEL: usize = 32;
```
Comment (buffalu) [inline ledger/src/bigtable_upload.rs:19]: note: i tried increasing this to 64 (running on 48 thread cpu) and it would lockup. don't really have an idea why this was happening yet. suspicious on the tokio runtime/something locking it up

**Comment by buffalu** [comment]
this line is sus: https://github.com/solana-labs/solana/blob/master/ledger/src/bigtable_upload.rs#L173

using blocking receiver and passing it to async thingy? cc @mvines

File: storage-bigtable/src/lib.rs [hunk 6]
```diff
@@ -790,21 +790,41 @@ impl LedgerStorage {
...
                     "tx-by-addr",
                     &tx_by_addr_cells,
                 )
-                .await?;
+                .await
+            }));
+        }
+
+        let results = futures::future::join_all(tasks).await;
+        let results: Vec<_> = results.into_iter().map(|r| r.unwrap()).collect();
```
Comment (t-nelson) [inline storage-bigtable/src/lib.rs:815]:
```suggestion
        let results: results.into_iter().map(|r| r).collect::<Result<Vec<_>>>()?;
```

Then I think you can clean up the stuff below, right?

File: ledger/src/bigtable_upload.rs (hunk 5 above)
Comment (0xEdgar) [reply ledger/src/bigtable_upload.rs:19]: with that many threads the locking could just be resource contention with the rest of the validato. my understanding is this upload service was originally supposed to run as just one background thread so would have a low resource impact

File: ledger/src/bigtable_upload.rs [hunk 7]
```diff
@@ -186,7 +196,12 @@ pub async fn upload_confirmed_blocks(
                 num_blocks -= 1;
                 None
             }
-            Some(confirmed_block) => Some(bigtable.upload_confirmed_block(slot, confirmed_block)),
+            Some(confirmed_block) => {
+                let bt = bigtable.clone();
+                Some(tokio::spawn(async move {
+                    bt.upload_confirmed_block(slot, confirmed_block).await
```
Comment (buffalu) [inline ledger/src/bigtable_upload.rs:202]: would be nice if there was a tokio receiver that let you clone the receiver and use multiple tasks to pop blocks off the queue and run in parallel; this thing could rip!

File: storage-bigtable/src/lib.rs (hunk 6 above)
Comment (buffalu) [reply storage-bigtable/src/lib.rs:815]: results is a Vec<Result<Result<usize>> where the first Result is a tokio::runtime::Result and the second one is a bigtable Result. 

i figure the join is _probably_ safe to unwrap, the bigtable result much more likely to fail

File: ledger/src/bigtable_upload.rs [hunk 8]
```diff
@@ -15,18 +15,32 @@ use {
     },
 };
 
-// Attempt to upload this many blocks in parallel
-const NUM_BLOCKS_TO_UPLOAD_IN_PARALLEL: usize = 32;
+pub struct ConfirmedBlockUploadConfig {
```
Comment (buffalu) [inline ledger/src/bigtable_upload.rs:18]: @t-nelson @CriesofCarrots lmk what you think of this

Comment (buffalu) [reply ledger/src/bigtable_upload.rs:18]: we're running this on our RPC server right now with no issues (yet) :)

File: ledger/src/bigtable_upload.rs (hunk 3 above)
Comment (buffalu) [reply ledger/src/bigtable_upload.rs]: added a config object, lmk what you think

File: ledger/src/bigtable_upload.rs (hunk 4 above)
Comment (buffalu) [reply ledger/src/bigtable_upload.rs]: fixed

File: storage-bigtable/src/lib.rs (hunk 6 above)
Comment (CriesofCarrots) [reply storage-bigtable/src/lib.rs:815]: I believe Trent is suggesting you use the `collect()` to find the first bigtable error, meaning you don't need the block on L825 below.
The bytes_written logic would have to come into the iterator, of course.

I'm torn on whether to unwrap the JoinError. Pondering that more.

File: ledger/src/bigtable_upload.rs [hunk 9]
```diff
@@ -15,18 +15,32 @@ use {
...
+    pub force_reupload: bool,
+    // number of blocks uploaded in parallel
```
Comment (CriesofCarrots) [inline ledger/src/bigtable_upload.rs:20]: This comment seems redundant with the field name. Is there anything distinct we want to say here?

File: ledger/src/bigtable_upload.rs [hunk 10]
```diff
@@ -15,18 +15,32 @@ use {
...
+    pub num_blocks_to_upload_in_parallel: usize,
+    // note: this shall always be greater than num_blocks_to_upload_in_parallel
```
Comment (CriesofCarrots) [inline ledger/src/bigtable_upload.rs:22]:
```suggestion
    // note: this must always be greater than num_blocks_to_upload_in_parallel
```
?

File: storage-bigtable/Cargo.toml [hunk 11]
```diff
@@ -27,6 +28,7 @@ solana-sdk = { path = "../sdk", version = "=1.11.0" }
 solana-storage-proto = { path = "../storage-proto", version = "=1.11.0" }
 solana-transaction-status = { path = "../transaction-status", version = "=1.11.0" }
 thiserror = "1.0"
+tokio = "1.17.0"
```
Comment (CriesofCarrots) [inline storage-bigtable/Cargo.toml:31]:
```suggestion
tokio = "1"
```

**Review by CriesofCarrots — COMMENTED** [review]
What kind of upload speeds are you seeing with this changeset?

**Comment by buffalu** [comment]
> What kind of upload speeds are you seeing with this changeset?

i think i was seeing ~200mbit/s, but can't remember specifics. i can test it again by disabling bigtable uploads on our node, run for ~1-2 hours, stop it, then use the ledger-tool upload command. lmk if that's something you're interested in!

we're running on 25gbps server so we have plenty of bandwidth left over. if we could have a threadpool that's constantly popping items off a queue and running i think we could get that number way up (as opposed to running NUM_PARALLEL uploads and waiting for them all to complete before popping more off.

> Why did you change your mind about the blockstore threads?

@t-nelson mentioned something about number of CPUs. I feel like that's probably better, but tbh don't have any super strong feelings about it. more is probably better tho, esp. if you're trying to catchup! during normal operation im seeing it keep up just fine and upload 2-3 blocks at a time with tip

**Comment by buffalu** [comment]
just wrote a script to check our bigtable highest slot against yours:
```
[2022-04-22T01:02:42Z INFO  jito_bigtable] destination bigtable 12496 slots ahead of source (source=130764747 dest=130777243)
```

File: storage-bigtable/src/lib.rs (hunk 6 above)
Comment (t-nelson) [reply storage-bigtable/src/lib.rs:815]: ah forget about the `JoinErr` :hankey: 

we don't actually need to spawn for each future, do we?  Just collect them all, then we can spawn once for `join_all`?  That'd give us a `Result<Iterator<bigtable::Result<usize>>>` that we can `result?.collect::<Result<Vec<usize>>()` on

Comment (t-nelson) [reply storage-bigtable/src/lib.rs:815]: n/m, we're already async.  don't think we need to spawn at all?

Comment (buffalu) [reply storage-bigtable/src/lib.rs:815]: most of the speed up actually came from the spawn!

Comment (buffalu) [reply storage-bigtable/src/lib.rs:815]: idk exactly how join all works but guessing they were running on a single thread?

Comment (CriesofCarrots) [reply storage-bigtable/src/lib.rs:815]: @t-nelson , what does the sentence with the emoji mean, when it's at home?

Comment (t-nelson) [reply storage-bigtable/src/lib.rs:815]: :fax: -> fax -> facts

Comment (CriesofCarrots) [reply storage-bigtable/src/lib.rs:815]: So: "I confirmed that we don't need to spawn since we're already async"?

Comment (t-nelson) [reply storage-bigtable/src/lib.rs:815]: > idk exactly how join all works but guessing they were running on a single thread?

I don't think so.  I used this to test my theory, https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a012d7c439412a36c05acc696ccfd601

Comment (t-nelson) [reply storage-bigtable/src/lib.rs:815]: > So: "I confirmed that we don't need to spawn since we're already async"?

i see that you are fluent in late-night-trent! yep!

**Comment by buffalu** [comment]
git commit: 00c5ec9673f9268751149c46773d36ba9ed9592b

num_cpus: 48
uses 12 threads (num_cpus / 4) for blockstore + upload
~100ms per block in the upload
```
[2022-04-22T18:27:42.137475474Z INFO  solana_ledger::bigtable_upload] Upload took 1.5s for 12 blocks
[2022-04-22T18:27:43.210726324Z INFO  solana_ledger::bigtable_upload] Upload took 1.1s for 12 blocks
[2022-04-22T18:27:44.252376864Z INFO  solana_ledger::bigtable_upload] Upload took 1.0s for 9 blocks
```
Screen Shot 2022-04-22 at 1 27 41 PM


num_cpus: 48
uses 24 threads (num_cpus / 2) for blockstore + upload
~70ms per block in the upload
```
[2022-04-22T18:22:19.896359960Z INFO  solana_ledger::bigtable_upload] Upload took 1.6s for 22 blocks
[2022-04-22T18:22:21.364399516Z INFO  solana_ledger::bigtable_upload] Upload took 1.5s for 21 blocks
[2022-04-22T18:22:45.426086757Z INFO  solana_ledger::bigtable_upload] Upload took 1.6s for 23 blocks
```
Screen Shot 2022-04-22 at 1 22 37 PM

**Comment by buffalu** [comment]
git commit 62a40f17cf51eda52a848207fbe11913e9c8017a 
(no tokio::spawn)
~200ms per block
```
[2022-04-22T18:51:03.401708893Z INFO  solana_ledger::bigtable_upload] Upload took 2.9s for 12 blocks
[2022-04-22T18:51:05.920438615Z INFO  solana_ledger::bigtable_upload] Upload took 2.5s for 12 blocks
[2022-04-22T18:51:08.139508376Z INFO  solana_ledger::bigtable_upload] Upload took 2.2s for 11 blocks
[2022-04-22T18:51:09.913070966Z INFO  solana_ledger::bigtable_upload] Upload took 1.8s for 10 blocks
[2022-04-22T18:51:11.336636226Z INFO  solana_ledger::bigtable_upload] Upload took 1.4s for 8 blocks
```
Screen Shot 2022-04-22 at 1 50 57 PM

File: storage-bigtable/src/lib.rs (hunk 6 above)
Comment (buffalu) [reply storage-bigtable/src/lib.rs:815]: copying from discord DMs:

i think if we spawned in your code we wouldn't see the 0,1,2,etc. print out in order, doing that now
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=85e14e6f14930884483eeda7aa64980b

suggests to me the old code is running async but single threaded, where this code running async and multithreaded.

another reason i hate async 👿

**Comment by buffalu** [comment]
git commit 0d797e2fff9554fbad6b810104119d7c578283fa (master)
```
[2022-04-22T20:01:47.160736902Z INFO  solana_ledger::bigtable_upload] Upload took 6.8s for 2 blocks
[2022-04-22T20:02:09.115707196Z INFO  solana_ledger::bigtable_upload] Upload took 22.0s for 32 blocks
[2022-04-22T20:02:26.023151730Z INFO  solana_ledger::bigtable_upload] Upload took 16.9s for 32 blocks
[2022-04-22T20:03:06.173026343Z INFO  solana_ledger::bigtable_upload] Upload took 19.9s for 32 blocks
```
Screen Shot 2022-04-22 at 3 01 54 PM

File: storage-bigtable/src/lib.rs (hunk 6 above)
Comment (t-nelson) [reply storage-bigtable/src/lib.rs:815]: right, but being as it's IO bound ops, the value of multi-threaded vs. single-thread + multiplexed should be nil so long as everyone is out of each other's way. i suspect that any gains from multi-threading are hiding some other issue

Comment (buffalu) [reply storage-bigtable/src/lib.rs:815]: everytime we upload we try 3 different compression methods and choose the best. there's our cpu bound

**Comment by buffalu** [comment]
in summary, seems like uploading blocks is network + cpu bound. 

haven't profiled, but guessing attempting to compressing each item that gets uploaded to bigtable 3 different ways is expensive 😆

**Comment by buffalu** [comment]
@CriesofCarrots are you running this on your warehouse nodes now? looks like they're caught up :)

**Comment by CriesofCarrots** [comment]
> are you running this on your warehouse nodes now? looks like they're caught up :)

No, I'm not sure why we're caught up suddenly 🤷‍♀️ (although Joe might know) 
Btw, I'm probably going to sneak in another change in this area that's going to force some rebasing here. Apologies in advance!

**Comment by buffalu** [comment]
> > are you running this on your warehouse nodes now? looks like they're caught up :)
> 
> No, I'm not sure why we're caught up suddenly 🤷‍♀️ (although Joe might know) Btw, I'm probably going to sneak in another change in this area that's going to force some rebasing here. Apologies in advance!

all good! feel free to ping me here or discord if you need anything! just rebased!

**Comment by CriesofCarrots** [comment]
Hey @buffalu , sorry for the delay here. I'm hoping to wrap up my changes in this area today/tomorrow. I played around with some of your changes on top, and it seems like the new spawns do make a huge difference. But reading the blocks from Blockstore was never a limiting factor on our warehouse nodes, so I think I'd like to first try a smaller changeset here without the multiple Blockstore threads. What do you think?
No need to take any action here yet; I'll ping you when my stuff is merged. Just wanted to start chatting about slimming this a little.

**Comment by buffalu** [comment]
> Hey @buffalu , sorry for the delay here. I'm hoping to wrap up my changes in this area today/tomorrow. I played around with some of your changes on top, and it seems like the new spawns do make a huge difference. But reading the blocks from Blockstore was never a limiting factor on our warehouse nodes, so I think I'd like to first try a smaller changeset here without the multiple Blockstore threads. What do you think? No need to take any action here yet; I'll ping you when my stuff is merged. Just wanted to start chatting about slimming this a little.

from what i remember i do think i was seeing blockstore take ~100ms per block so it might be worth checking to make sure it's not going to be a limiting factor if someone wants to crank it up more (ie using ledger-tool/custom number of cpus in the config).

**Comment by buffalu** [comment]
> @buffalu , I'm finally done mucking around in the bigtable_upload files. Are you still willing to rebase this? I see your point about the Blockstore threads. It's never the bottleneck on our nodes, but just talked to a partner blocked by it.

yeah! gimme a few hours, will get to later after meetings 😢

**Comment by buffalu** [comment]
@CriesofCarrots can you elaborate what you mean by "It's never the bottleneck on our nodes, but just talked to a partner blocked by it."

are they running with the new tokio::spawn but single blockstore thread? or just in general bigtable upload being slow

**Comment by CriesofCarrots** [comment]
> are they running with the new tokio::spawn but single blockstore thread? or just in general bigtable upload being slow

They are running with vanilla v1.9 (don't recall which patch), and it was clear from the timings that the upload thread was sitting idle waiting for blockstore reads

**Comment by buffalu** [comment]
ok should be good. caught a potential unwrap error + fixed that. lmk if anything else sticks out

File: storage-bigtable/Cargo.toml [hunk 12]
```diff
@@ -29,6 +30,7 @@ solana-transaction-status = { path = "../transaction-status", version = "=1.11.0
 thiserror = "1.0"
 tonic = { version = "0.7.2", features = ["tls", "transport"] }
 zstd = "0.11.2"
+tokio = "1"
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply storage-bigtable/Cargo.toml:33]: Also A-Z this, please

**Review by CriesofCarrots — COMMENTED** [review]
I started reviewing this, but looks like we lost your changes to use num_cpus as basis for num_blocks_to_upload_in_parallel. Can you restore that? 🙏

**Comment by buffalu** [comment]
> I started reviewing this, but looks like we lost your changes to use num_cpus as basis for num_blocks_to_upload_in_parallel. Can you restore that? 🙏

yeah, assumed we wanted to use your config already. do you think it makes sense for num blockstore threads = num sending threads = num_blocks_to_upload_in_parallel?

**Comment by CriesofCarrots** [comment]
> yeah, assumed we wanted to use your config already.

Oh gotcha. Sorry if that was confusing. I just set up relative values.
I still think it makes sense to base the defaults on num_cpus; probably half num_cpus for num_blocks_to_upload_in_parallel...? Was that what you had before?

>  do you think it makes sense for num blockstore threads = num sending threads = num_blocks_to_upload_in_parallel?

Yes, that makes sense to me

**Comment by buffalu** [comment]
this prob doesn't make sense for this PR, but we also noticed using separate LedgerStorage instances can massively speed things up too. we can get ~500-1k blocks per second reading with this change across 64 threads and 500 slot request sizes.

cloning causes them to use the same channel and TCP connection, if you create a new instance for each thread they'd each have their own connection + channel

**Comment by CriesofCarrots** [comment]
> we also noticed using separate LedgerStorage instances can massively speed things up too

Yeah, let's look at that separately

File: storage-bigtable/src/lib.rs (hunk 6 above)
Comment (CriesofCarrots) [inline storage-bigtable/src/lib.rs:815]: Let's map the tokio Error to something, just to be safe. Can you add an `Error` variant?
Also, can we rewrite this whole block to only iterate once?
Something like:
```
let mut bytes_written = 0;
let mut maybe_first_err: Option<Error> = None;
for result in results {
    match result {
        Err(err) => if maybe_first_err.is_none() {
            maybe_first_err = Some(Error::TokioError(err));
        }
        Ok(Err(err)) => if maybe_first_err.is_none() {
            maybe_first_err = Some(Error::BigTableError(err));
        }
        Ok(Ok(bytes)) => {
            bytes_written += bytes;
        }
    }
}
if let Some(err) = maybe_first_err {
    return Err(err);
}
```

**Comment by CriesofCarrots** [comment]
One additional request: when you pin `tokio`, can you also please rollback all the Cargo.lock version bumps? Would prefer to consider those separately.
I think the only Cargo.lock changes needed should be adding tokio and futures to the solana-bigtable crate.

File: storage-bigtable/Cargo.toml [hunk 13]
```diff
@@ -18,6 +18,7 @@ flate2 = "1.0.23"
 goauth = "0.12.0"
 log = "0.4.17"
 prost = "0.10.3"
+futures = "0.3.21"
```
[the remark this answers is not in the document]
Comment (buffalu) [reply storage-bigtable/Cargo.toml]: fixed

File: storage-bigtable/Cargo.toml (hunk 12 above)
[the remark this answers is not in the document]
Comment (buffalu) [reply storage-bigtable/Cargo.toml]: fixed

File: ledger/src/bigtable_upload.rs [hunk 14]
```diff
@@ -207,12 +244,26 @@ pub async fn upload_confirmed_blocks(
...
+                error!(
+                    "upload_confirmed_block() failed: {:?}",
+                    result.as_ref().err()
+                );
+                failures += 1;
+            } else if result.as_ref().unwrap().is_err() {
+                error!(
+                    "upload_confirmed_block() failed: {:?}",
+                    result.unwrap().err()
+                );
```
[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs]: fixed

File: storage-bigtable/src/lib.rs [hunk 15]
```diff
@@ -790,21 +795,51 @@ impl LedgerStorage {
...
+                conn.put_protobuf_cells_with_retry::<tx_by_addr::TransactionByAddr>(
                     "tx-by-addr",
                     &tx_by_addr_cells,
                 )
-                .await?;
+                .await
+            }));
+        }
+
+        let mut bytes_written = 0;
```
Comment (CriesofCarrots) [inline storage-bigtable/src/lib.rs:819]: Oops, sorry, bad reviewer. We already have this defined on L740, so can remove here.

Comment (buffalu) [reply storage-bigtable/src/lib.rs:819]: oh yeah good catch, totally missed that :)

**Review by CriesofCarrots — COMMENTED** [review]
I have this running on two testnet warehouse nodes, and it's working great! I noticed a couple logging things (see comments).
Otherwise, just the Cargo.lock reverts, and I think that's it from me!

File: ledger/src/bigtable_upload.rs [hunk 16]
```diff
@@ -147,42 +147,79 @@ pub async fn upload_confirmed_blocks(
...
+                            if num_blocks_uploaded > 0
+                                && num_blocks_uploaded % num_to_upload_in_parallel == 0
+                            {
+                                info!(
+                                    "blockstore_upload thread {} processed {} blocks (elapsed {:.2})",
+                                    i,
+                                    num_blocks_uploaded,
+                                    start.elapsed().as_secs_f32()
+                                );
+                            }
```
[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs:211]: will remove

File: ledger/src/bigtable_upload.rs [hunk 17]
```diff
@@ -147,42 +147,79 @@ pub async fn upload_confirmed_blocks(
...
+                        let elapsed = start.elapsed();
                         info!(
-                            "{}% of blocks processed ({}/{})",
-                            i * 100 / blocks_to_upload.len(),
+                            "blockstore_upload thread {} took {:.2} to load {} blocks",
                             i,
-                            blocks_to_upload.len()
+                            elapsed.as_secs_f32(),
+                            num_blocks_uploaded
                         );
```
[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs:219]: wdyt on returning elapsed, running a join (we don't join these threads rn) and summing all the durations together returned by the join handle?

[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs:219]: could also have an atomic and sum into that inside the thread

[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply ledger/src/bigtable_upload.rs:219]: > wdyt on returning elapsed, running a join (we don't join these threads rn) and summing all the durations together returned by the join handle?

Yeah, I was thinking join also. If we join, we could construct a Measure before and then just log elapsed after the join, I think?

[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs:219]: Wfm. Also having a hard time getting rid of those cargo.lock changes, did cargo clean and removed that lock file and still getting issues. I’ll try checkout master .lock and rebuild and see if that fixes it

File: ledger/src/bigtable_upload.rs [hunk 18]
```diff
@@ -223,6 +257,35 @@ pub async fn upload_confirmed_blocks(
 
     measure.stop();
     info!("{}", measure);
+
+    let blockstore_results: Vec<std::thread::Result<BlockstoreLoadStats>> =
```
Comment (buffalu) [inline ledger/src/bigtable_upload.rs:261]: cc @CriesofCarrots lmk what you think of this

File: ledger/src/bigtable_upload.rs [hunk 19]
```diff
@@ -223,6 +257,35 @@ pub async fn upload_confirmed_blocks(
...
+
+    let mut blockstore_num_blocks_read = 0;
+    let mut blockstore_elapsed = Duration::default();
+    let mut blockstore_errors = 0;
+
+    for r in blockstore_results {
+        match r {
+            Ok(stats) => {
+                blockstore_num_blocks_read += stats.num_blocks_read;
+                blockstore_elapsed += stats.elapsed;
```
[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs:272]: oh yeah, this is what i get for rushing lol

[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs:272]: wdyt about taking the max stats.elapsed? theres a chance the loader thread could be joinable mid-thread upload_confirmed_block thread. if you assume all the blockstore threads start at the same time, then the max elapsed of all the threads would be equal to wallclock time?

depends on how accurate you'd like to be :)

File: ledger/src/bigtable_upload.rs [hunk 20]
```diff
@@ -223,6 +257,35 @@ pub async fn upload_confirmed_blocks(
 
     measure.stop();
     info!("{}", measure);
+
+    let blockstore_results: Vec<std::thread::Result<BlockstoreLoadStats>> =
+        loader_threads.into_iter().map(|t| t.join()).collect();
```
Comment (t-nelson) [inline ledger/src/bigtable_upload.rs:262]: I think the `collect()` here is superfluous?  Should be able to just update the stats directly from the iterator

File: storage-bigtable/src/lib.rs [hunk 21]
```diff
@@ -54,6 +56,9 @@ pub enum Error {
 
     #[error("Signature not found")]
     SignatureNotFound,
+
+    #[error("tokio error")]
+    TokioError(JoinError),
```
Comment (t-nelson) [inline storage-bigtable/src/lib.rs:61]:
```suggestion
    TokioJoinError(JoinError),
```

**Review by t-nelson — APPROVED** [review]
lgtm. just some nits

File: ledger/src/bigtable_upload.rs (hunk 20 above)
Comment (buffalu) [reply ledger/src/bigtable_upload.rs]: fixed

File: ledger/src/bigtable_upload.rs [hunk 22]
```diff
@@ -207,12 +233,20 @@ pub async fn upload_confirmed_blocks(
...
         });
 
         for result in futures::future::join_all(uploads).await {
-            if result.is_err() {
-                error!("upload_confirmed_block() failed: {:?}", result.err());
+            if let Err(err) = result {
+                error!("upload_confirmed_block() failed: {:?}", err);
+                failures += 1;
+            } else if let Err(err) = result.unwrap() {
+                error!("upload_confirmed_block() failed: {:?}", err);
```
[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs]: fixed

File: ledger/src/bigtable_upload.rs [hunk 23]
```diff
@@ -207,12 +233,20 @@ pub async fn upload_confirmed_blocks(
...
+                    bt.upload_confirmed_block(slot, confirmed_block).await
+                }))
+            }
         });
 
         for result in futures::future::join_all(uploads).await {
-            if result.is_err() {
-                error!("upload_confirmed_block() failed: {:?}", result.err());
+            if let Err(err) = result {
+                error!("upload_confirmed_block() failed: {:?}", err);
```
[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs]: fixed

File: ledger/src/bigtable_upload.rs [hunk 24]
```diff
@@ -147,42 +152,63 @@ pub async fn upload_confirmed_blocks(
...
+                            if exit.load(Ordering::Relaxed) {
+                                break;
+                            }
+
+                            let slot = match slot_receiver.recv() {
+                                Ok(slot) => slot,
+                                Err(_) => {
+                                    break;
+                                }
+                            };
```
[the remark this answers is not in the document]
Comment (buffalu) [reply ledger/src/bigtable_upload.rs]: fixed

File: storage-bigtable/src/lib.rs (hunk 21 above)
Comment (buffalu) [reply storage-bigtable/src/lib.rs]: fixed

File: ledger/src/bigtable_upload.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply ledger/src/bigtable_upload.rs]: Checking out master .lock and rebuilding works on my end. Can you give it a go?

[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply ledger/src/bigtable_upload.rs]: \o/

**Review by CriesofCarrots — APPROVED** [review]
Thanks for all the iterations on this @buffalu !

**Comment by CriesofCarrots** [comment]
Merging on red; downstream-anchor-projects has been removed on master (saving CI the load of rebasing this)
